File: anyflix-backend/lib/models/tmdb.py
"""TMDB (The Movie Database) models."""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class TMDBGenre(BaseModel):
    """TMDB genre model."""

    id: int
    name: str


class TMDBProductionCompany(BaseModel):
    """TMDB production company model."""

    id: int
    logo_path: str | None = None
    name: str
    origin_country: str


class TMDBProductionCountry(BaseModel):
    """TMDB production country model."""

    iso_3166_1: str
    name: str


class TMDBSpokenLanguage(BaseModel):
    """TMDB spoken language model."""

    english_name: str
    iso_639_1: str
    name: str


class TMDBVideoType(str, Enum):
    """TMDB video type enumeration."""

    TRAILER = "Trailer"
    CLIP = "Clip"
    TEASER = "Teaser"
    OPENING_CREDITS = "Opening Credits"
    ENDING_CREDITS = "Ending Credits"


class TMDBVideoSite(str, Enum):
    """TMDB video site enumeration."""

    YOUTUBE = "YouTube"


class TMDBVideo(BaseModel):
    """TMDB video model."""

    iso_639_1: str
    iso_3166_1: str
    name: str
    key: str
    site: str | TMDBVideoSite
    size: int
    type: TMDBVideoType | str | None = None
    official: bool
    published_at: str
    id: str


class TMDBVideoResult(BaseModel):
    """TMDB video result model."""

    results: list[TMDBVideo] = Field(default_factory=list)


class TMDBImage(BaseModel):
    """TMDB image model."""

    aspect_ratio: float
    height: int
    iso_639_1: str | None = None
    file_path: str
    vote_average: float
    vote_count: int
    width: int


class TMDBImages(BaseModel):
    """TMDB images collection."""

    backdrops: list[TMDBImage] = Field(default_factory=list)
    logos: list[TMDBImage] = Field(default_factory=list)
    posters: list[TMDBImage] = Field(default_factory=list)


class TMDBExternalIds(BaseModel):
    """TMDB external IDs."""

    imdb_id: str | None = None
    freebase_mid: str | None = None
    freebase_id: str | None = None
    tvdb_id: int | None = None
    tvrage_id: int | None = None
    wikidata_id: str | None = None
    facebook_id: str | None = None
    instagram_id: str | None = None
    twitter_id: str | None = None


class TMDBCreatedBy(BaseModel):
    """TMDB created by person."""

    id: int
    credit_id: str
    name: str
    gender: int
    profile_path: str | None = None


class TMDBNetwork(BaseModel):
    """TMDB network model."""

    id: int
    logo_path: str | None = None
    name: str
    origin_country: str


class TMDBSeason(BaseModel):
    """TMDB season model."""

    air_date: str | None = None
    episode_count: int
    id: int
    name: str
    overview: str
    poster_path: str | None = None
    season_number: int
    vote_average: float


class TMDBEpisode(BaseModel):
    """TMDB episode model."""

    id: int
    name: str
    overview: str
    vote_average: float
    vote_count: int
    air_date: str | None = None
    episode_number: int
    episode_type: str
    production_code: str
    runtime: int | None = None
    season_number: int
    show_id: int
    still_path: str | None = None


class TMDBMovieDetail(BaseModel):
    """TMDB movie detail model."""

    adult: bool
    backdrop_path: str | None = None
    belongs_to_collection: dict[str, Any] | None = None
    budget: int
    genres: list[TMDBGenre] = Field(default_factory=list)
    homepage: str | None = None
    id: int
    imdb_id: str | None = None
    original_language: str
    original_title: str
    overview: str | None = None
    popularity: float
    poster_path: str | None = None
    production_companies: list[TMDBProductionCompany] = Field(default_factory=list)
    production_countries: list[TMDBProductionCountry] = Field(default_factory=list)
    release_date: str | None = None
    revenue: int
    runtime: int | None = None
    spoken_languages: list[TMDBSpokenLanguage] = Field(default_factory=list)
    status: str
    tagline: str | None = None
    title: str
    video: bool
    vote_average: float
    vote_count: int

    # Additional fields when requested
    videos: TMDBVideoResult | None = None
    images: TMDBImages | None = None
    external_ids: TMDBExternalIds | None = None


class TMDBTVDetail(BaseModel):
    """TMDB TV show detail model."""

    adult: bool
    backdrop_path: str | None = None
    created_by: list[TMDBCreatedBy] = Field(default_factory=list)
    episode_run_time: list[int] = Field(default_factory=list)
    first_air_date: str | None = None
    genres: list[TMDBGenre] = Field(default_factory=list)
    homepage: str
    id: int
    in_production: bool
    languages: list[str] = Field(default_factory=list)
    last_air_date: str | None = None
    last_episode_to_air: TMDBEpisode | None = None
    name: str
    networks: list[TMDBNetwork] = Field(default_factory=list)
    next_episode_to_air: TMDBEpisode | None = None
    number_of_episodes: int
    number_of_seasons: int
    origin_country: list[str] = Field(default_factory=list)
    original_language: str
    original_name: str
    overview: str
    popularity: float
    poster_path: str | None = None
    production_companies: list[TMDBProductionCompany] = Field(default_factory=list)
    production_countries: list[TMDBProductionCountry] = Field(default_factory=list)
    seasons: list[TMDBSeason] = Field(default_factory=list)
    spoken_languages: list[TMDBSpokenLanguage] = Field(default_factory=list)
    status: str
    tagline: str
    type: str
    vote_average: float
    vote_count: int

    # Additional fields when requested
    videos: TMDBVideoResult | None = None
    images: TMDBImages | None = None
    external_ids: TMDBExternalIds | None = None


class TMDBSearchResult(BaseModel):
    """TMDB search result model."""

    id: int
    media_type: str  # "movie", "tv", or "person"
    adult: bool | None = None
    backdrop_path: str | None = None
    poster_path: str | None = None
    popularity: float = 0.0
    vote_average: float | None = None
    vote_count: int | None = None
    overview: str | None = None
    genre_ids: list[int] = Field(default_factory=list)
    original_language: str | None = None

    # Movie specific fields
    title: str | None = None
    original_title: str | None = None
    release_date: str | None = None
    video: bool | None = None

    # TV specific fields
    name: str | None = None
    original_name: str | None = None
    first_air_date: str | None = None
    origin_country: list[str] = Field(default_factory=list)

    # Person specific fields (for when media_type is "person")
    profile_path: str | None = None
    known_for: list[dict[str, Any]] = Field(default_factory=list)
    known_for_department: str | None = None
    gender: int | None = None


class TMDBSearchResponse(BaseModel):
    """TMDB search response model."""

    page: int
    results: list[TMDBSearchResult] = Field(default_factory=list)
    total_pages: int
    total_results: int


class TMDBConfiguration(BaseModel):
    """TMDB configuration model."""

    images: dict[str, Any]
    change_keys: list[str] = Field(default_factory=list)


# TMDB Movie Genres
TMDB_MOVIE_GENRES = {
    28: "Action",
    12: "Adventure",
    16: "Animation",
    35: "Comedy",
    80: "Crime",
    99: "Documentary",
    18: "Drama",
    10751: "Family",
    14: "Fantasy",
    36: "History",
    27: "Horror",
    10402: "Music",
    9648: "Mystery",
    10749: "Romance",
    878: "Science Fiction",
    10770: "TV Movie",
    53: "Thriller",
    10752: "War",
    37: "Western",
}

# TMDB TV Genres
TMDB_TV_GENRES = {
    10759: "Action & Adventure",
    16: "Animation",
    35: "Comedy",
    80: "Crime",
    99: "Documentary",
    18: "Drama",
    10751: "Family",
    10762: "Kids",
    9648: "Mystery",
    10763: "News",
    10764: "Reality",
    10765: "Sci-Fi & Fantasy",
    10766: "Soap",
    10767: "Talk",
    10768: "War & Politics",
    37: "Western",
}

# Combined genre mapping for comprehensive lookup
TMDB_ALL_GENRES = {
    **TMDB_MOVIE_GENRES,
    **TMDB_TV_GENRES,
}


def get_genre_name(genre_id: int, media_type: str = "all") -> str | None:
    """
    Get genre name by ID.

    Args:
        genre_id: The TMDB genre ID
        media_type: Type of media ("movie", "tv", or "all")

    Returns:
        Genre name if found, None otherwise
    """
    if media_type == "movie":
        return TMDB_MOVIE_GENRES.get(genre_id)
    if media_type == "tv":
        return TMDB_TV_GENRES.get(genre_id)
    return TMDB_ALL_GENRES.get(genre_id)


def get_genres_by_ids(genre_ids: list[int], media_type: str = "all") -> list[str]:
    """
    Get genre names by list of IDs.

    Args:
        genre_ids: List of TMDB genre IDs
        media_type: Type of media ("movie", "tv", or "all")

    Returns:
        List of genre names (excludes None values)
    """
    genres = []
    for genre_id in genre_ids:
        genre_name = get_genre_name(genre_id, media_type)
        if genre_name:
            genres.append(genre_name.lower())
    return genres


def find_genre_id(genre_name: str, media_type: str = "all") -> int | None:
    """
    Find genre ID by name (case-insensitive).

    Args:
        genre_name: The genre name to search for
        media_type: Type of media ("movie", "tv", or "all")

    Returns:
        Genre ID if found, None otherwise
    """
    genre_name_lower = genre_name.lower()

    search_dict = TMDB_ALL_GENRES
    if media_type == "movie":
        search_dict = TMDB_MOVIE_GENRES
    elif media_type == "tv":
        search_dict = TMDB_TV_GENRES

    for genre_id, name in search_dict.items():
        if name.lower() == genre_name_lower:
            return genre_id

    return None

File: anyflix-player/src/types/schemas/tmdb.ts
/**
 * TMDB (The Movie Database) Zod schemas
 * Complete TMDB API data models with runtime validation
 */

import { z } from 'zod';

// =============================================
// TMDB Core Models
// =============================================

export const TMDBGenreSchema = z.object({
  id: z.number(),
  name: z.string()
});
export type TMDBGenre = z.infer<typeof TMDBGenreSchema>;

export const TMDBProductionCompanySchema = z.object({
  id: z.number(),
  logo_path: z.string().optional(),
  name: z.string(),
  origin_country: z.string()
});
export type TMDBProductionCompany = z.infer<typeof TMDBProductionCompanySchema>;

export const TMDBProductionCountrySchema = z.object({
  iso_3166_1: z.string(),
  name: z.string()
});
export type TMDBProductionCountry = z.infer<typeof TMDBProductionCountrySchema>;

export const TMDBSpokenLanguageSchema = z.object({
  english_name: z.string(),
  iso_639_1: z.string(),
  name: z.string()
});
export type TMDBSpokenLanguage = z.infer<typeof TMDBSpokenLanguageSchema>;

export const TMDBVideoSchema = z.object({
  iso_639_1: z.string(),
  iso_3166_1: z.string(),
  name: z.string(),
  key: z.string(),
  site: z.string(),
  size: z.number(),
  type: z.string(),
  official: z.boolean(),
  published_at: z.string(),
  id: z.string()
});
export type TMDBVideo = z.infer<typeof TMDBVideoSchema>;

export const TMDBImageSchema = z.object({
  aspect_ratio: z.number(),
  height: z.number(),
  iso_639_1: z.string().optional(),
  file_path: z.string(),
  vote_average: z.number(),
  vote_count: z.number(),
  width: z.number()
});
export type TMDBImage = z.infer<typeof TMDBImageSchema>;

export const TMDBImagesSchema = z.object({
  backdrops: z.array(TMDBImageSchema).default([]),
  logos: z.array(TMDBImageSchema).default([]),
  posters: z.array(TMDBImageSchema).default([])
});
export type TMDBImages = z.infer<typeof TMDBImagesSchema>;

export const TMDBExternalIdsSchema = z.object({
  imdb_id: z.string().optional(),
  freebase_mid: z.string().optional(),
  freebase_id: z.string().optional(),
  tvdb_id: z.number().optional(),
  tvrage_id: z.number().optional(),
  wikidata_id: z.string().optional(),
  facebook_id: z.string().optional(),
  instagram_id: z.string().optional(),
  twitter_id: z.string().optional()
});
export type TMDBExternalIds = z.infer<typeof TMDBExternalIdsSchema>;

// =============================================
// TMDB TV-specific Models
// =============================================

export const TMDBCreatedBySchema = z.object({
  id: z.number(),
  credit_id: z.string(),
  name: z.string(),
  gender: z.number(),
  profile_path: z.string().optional()
});
export type TMDBCreatedBy = z.infer<typeof TMDBCreatedBySchema>;

export const TMDBNetworkSchema = z.object({
  id: z.number(),
  logo_path: z.string().optional(),
  name: z.string(),
  origin_country: z.string()
});
export type TMDBNetwork = z.infer<typeof TMDBNetworkSchema>;

export const TMDBSeasonSchema = z.object({
  air_date: z.string().optional(),
  episode_count: z.number(),
  id: z.number(),
  name: z.string(),
  overview: z.string(),
  poster_path: z.string().optional(),
  season_number: z.number(),
  vote_average: z.number()
});
export type TMDBSeason = z.infer<typeof TMDBSeasonSchema>;

export const TMDBEpisodeSchema = z.object({
  id: z.number(),
  name: z.string(),
  overview: z.string(),
  vote_average: z.number(),
  vote_count: z.number(),
  air_date: z.string().optional(),
  episode_number: z.number(),
  episode_type: z.string(),
  production_code: z.string(),
  runtime: z.number().optional(),
  season_number: z.number(),
  show_id: z.number(),
  still_path: z.string().optional()
});
export type TMDBEpisode = z.infer<typeof TMDBEpisodeSchema>;

// =============================================
// TMDB Detail Models
// =============================================

export const TMDBMovieDetailSchema = z.object({
  adult: z.boolean(),
  backdrop_path: z.string().optional(),
  belongs_to_collection: z.record(z.any()).optional(),
  budget: z.number(),
  genres: z.array(TMDBGenreSchema).default([]),
  homepage: z.string().optional(),
  id: z.number(),
  imdb_id: z.string().optional(),
  original_language: z.string(),
  original_title: z.string(),
  overview: z.string().optional(),
  popularity: z.number(),
  poster_path: z.string().optional(),
  production_companies: z.array(TMDBProductionCompanySchema).default([]),
  production_countries: z.array(TMDBProductionCountrySchema).default([]),
  release_date: z.string().optional(),
  revenue: z.number(),
  runtime: z.number().optional(),
  spoken_languages: z.array(TMDBSpokenLanguageSchema).default([]),
  status: z.string(),
  tagline: z.string().optional(),
  title: z.string(),
  video: z.boolean(),
  vote_average: z.number(),
  vote_count: z.number(),
  // Additional fields when requested
  videos: z.record(z.array(TMDBVideoSchema)).optional(),
  images: TMDBImagesSchema.optional(),
  external_ids: TMDBExternalIdsSchema.optional()
});
export type TMDBMovieDetail = z.infer<typeof TMDBMovieDetailSchema>;

export const TMDBTVDetailSchema = z.object({
  adult: z.boolean(),
  backdrop_path: z.string().optional(),
  created_by: z.array(TMDBCreatedBySchema).default([]),
  episode_run_time: z.array(z.number()).default([]),
  first_air_date: z.string().optional(),
  genres: z.array(TMDBGenreSchema).default([]),
  homepage: z.string(),
  id: z.number(),
  in_production: z.boolean(),
  languages: z.array(z.string()).default([]),
  last_air_date: z.string().optional(),
  last_episode_to_air: TMDBEpisodeSchema.optional(),
  name: z.string(),
  networks: z.array(TMDBNetworkSchema).default([]),
  next_episode_to_air: TMDBEpisodeSchema.optional(),
  number_of_episodes: z.number(),
  number_of_seasons: z.number(),
  origin_country: z.array(z.string()).default([]),
  original_language: z.string(),
  original_name: z.string(),
  overview: z.string(),
  popularity: z.number(),
  poster_path: z.string().optional(),
  production_companies: z.array(TMDBProductionCompanySchema).default([]),
  production_countries: z.array(TMDBProductionCountrySchema).default([]),
  seasons: z.array(TMDBSeasonSchema).default([]),
  spoken_languages: z.array(TMDBSpokenLanguageSchema).default([]),
  status: z.string(),
  tagline: z.string(),
  type: z.string(),
  vote_average: z.number(),
  vote_count: z.number(),
  // Additional fields when requested
  videos: z.record(z.array(TMDBVideoSchema)).optional(),
  images: TMDBImagesSchema.optional(),
  external_ids: TMDBExternalIdsSchema.optional()
});
export type TMDBTVDetail = z.infer<typeof TMDBTVDetailSchema>;

// =============================================
// TMDB Search Models
// =============================================

export const TMDBSearchResultSchema = z.object({
  id: z.number(),
  media_type: z.string(),
  adult: z.boolean().optional(),
  backdrop_path: z.string().optional(),
  poster_path: z.string().optional(),
  popularity: z.number().default(0),
  vote_average: z.number().optional(),
  vote_count: z.number().optional(),
  overview: z.string().optional(),
  genre_ids: z.array(z.number()).default([]),
  original_language: z.string().optional(),
  // Movie specific fields
  title: z.string().optional(),
  original_title: z.string().optional(),
  release_date: z.string().optional(),
  video: z.boolean().optional(),
  // TV specific fields
  name: z.string().optional(),
  original_name: z.string().optional(),
  first_air_date: z.string().optional(),
  origin_country: z.array(z.string()).default([]),
  // Person specific fields
  profile_path: z.string().optional(),
  known_for: z.array(z.record(z.any())).default([]),
  known_for_department: z.string().optional(),
  gender: z.number().optional()
});
export type TMDBSearchResult = z.infer<typeof TMDBSearchResultSchema>;

export const TMDBSearchResponseSchema = z.object({
  page: z.number(),
  results: z.array(TMDBSearchResultSchema).default([]),
  total_pages: z.number(),
  total_results: z.number()
});
export type TMDBSearchResponse = z.infer<typeof TMDBSearchResponseSchema>;

// =============================================
// TMDB Configuration
// =============================================

export const TMDBConfigurationSchema = z.object({
  images: z.record(z.any()),
  change_keys: z.array(z.string()).default([])
});
export type TMDBConfiguration = z.infer<typeof TMDBConfigurationSchema>;
